src/DynamoCore/Scheduler/UpdateRenderPackageAsyncTask.cs

```diff
@@ -212,104 +280,173 @@ protected override void HandleTaskExecutionCore()
+        {
+            var packageSupplement = package as IRenderPackageSupplement;
+            
+            //Use legacy slow path if package does not implement IRenderPackageSupplement
+            if (packageSupplement == null)
+            {
+                LegacyPackageEnsureColorExistsPerVertex(package, previousPointVertexCount, previousLineVertexCount, previousMeshVertexCount);
+                return;
+            }
+            
+            if (package.PointVertexCount > previousPointVertexCount)
+            {
+                var count = package.PointVertexCount - packageSupplement.PointVertexColorCount;
+                if (count > 0)
+                {
+                    packageSupplement.AppendPointVertexColorRange(CreateColorByteArrayOfSize(count, DefR, DefG, DefB, DefA));
+                }
+                
+            }
 
-                    // The default color coming from the geometry library for
-                    // curves is 255,255,255,255 (White). Because we want a default
-                    // color of 0,0,0,255 (Black), we adjust the color components here.
-                    if (graphicItem is Curve || graphicItem is Surface || graphicItem is Solid || graphicItem is Point)
-                    {
-                        if (package.LineVertexCount > 0 && !package.LineStripVertexColors.Any())
-                        {
-                            package.ApplyLineVertexColors(CreateColorByteArrayOfSize(package.LineVertexCount, DefR, DefG, DefB, DefA));
-                        }
+            if (package.LineVertexCount > previousLineVertexCount)
+            {
+                var count = package.LineVertexCount - packageSupplement.LineVertexColorCount;
+                if (count > 0)
+                {
+                    packageSupplement.AppendLineVertexColorRange(CreateColorByteArrayOfSize(count, DefR, DefG, DefB, DefA));
+                }
+            }
 
-                        if (package.PointVertexCount > 0 && !package.PointVertexColors.Any())
-                        {
-                            package.ApplyPointVertexColors(CreateColorByteArrayOfSize(package.PointVertexCount, DefR, DefG, DefB, DefA));
-                        }
-                    }
+            if (package.MeshVertexCount > previousMeshVertexCount)
+            {
+                var count = package.MeshVertexCount - packageSupplement.MeshVertexColorCount;
+                if (count > 0)
+                {
+                    packageSupplement.AppendMeshVertexColorRange(CreateColorByteArrayOfSize(count, DefR, DefG, DefB, DefA));
                 }
-                catch (Exception e)
+            }
+        }
+
+        private static void LegacyPackageEnsureColorExistsPerVertex(IRenderPackage pkg, int previousPointVertexCount, int previousLineVertexCount, int previousMeshVertexCount)
```

Comment:
So this method essentially does the same thing as `EnsureColorExistsPerVertex` but with the old APIs? I suppose this check is needed only now after the change to a single render package per node? If so, when will there be a case where the package will not be an `IRenderPackageSupplement` but there would still be a single package per node?

Reply:
Is this method needed for custom third party `IRenderPackage `types that do not implement `IRenderPackageSupplement` but will still be running Dynamo after these changes?

Reply:
Yes this is for third party implementation of `IRenderPackage`.  Our Helix and Default implementations in Dynamo  would add `IRenderPackageSuppliment`.  The color check is done regardless of render package per `IGraphicItem` or per node.  This was previously handled here -> https://github.com/DynamoDS/Dynamo/blob/master/src/DynamoCore/Scheduler/UpdateRenderPackageAsyncTask.cs#L290.  This just ensures color info is added to the RenderPackage before being passed to the Helix View